### PR TITLE
FINERACT-1840: Fee income is not realised if the loan got fully repaid

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

A Loan with a NSF fee is added (specific due date fee) and the loan got fully repaid on the same day the Fee income is not realised, the Accrual transaction is not added

[FINERACT-1840](https://issues.apache.org/jira/browse/FINERACT-1840)

- Repayment Schedule
Screenshot 2023-01-22 at 16 46 26

- Transactions
Screenshot 2023-01-22 at 16 46 17

- Another example
Screenshot 2023-01-26 at 0 31 48

Screenshot 2023-01-26 at 0 31 04

Screenshot 2023-01-26 at 0 30 50

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Review by adamsaghy — CHANGES_REQUESTED** [review]
We should not rework an existing business logic without a good reason and proper test coverage. 
Please discuss these changes further!

File: fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java [hunk 1]
```diff
@@ -345,4 +347,19 @@ private Set<Long> getNotifiableUserIds(Long officeId, String permission) {
         Collection<AppUser> usersWithPermission = users.stream().filter(aU -> aU.hasAnyPermission(permission, "ALL_FUNCTIONS")).toList();
         return usersWithPermission.stream().map(AppUser::getId).collect(toSet());
     }
+
+    private static class LoanStatusChangedListener implements BusinessEventListener<LoanStatusChangedBusinessEvent> {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java]: Please move this logic into LoanWritePlatformServiceJpaRepositoryImpl or create a new service which will take care of the logic. I reckon the NotificationDomainService is not the proper place!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 2]
```diff
@@ -3451,9 +3451,9 @@ private void handleLoanRepaymentInFull(final LocalDate transactionDate, final Lo
             }
         }
         if (isAllChargesPaid) {
+            this.closedOnDate = transactionDate;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3450]: Why did you move the date? If the reason was to make this date be part of the event, please move the actualMaturityDate as well to be consistent.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 3]
```diff
@@ -3490,26 +3490,47 @@ private void processIncomeAccrualTransactionOnLoanClosure() {
...
 
+    public void applyIncomeAccrualTransaction(LocalDate closedDate) {
+        if (isPeriodicAccrualAccountingEnabledOnLoanProduct()) {
+            ExternalId externalId = ExternalId.empty();
+            final boolean isExternalIdAutoGenerationEnabled = TemporaryConfigurationServiceContainer.isExternalIdAutoGenerationEnabled();
+            if (isExternalIdAutoGenerationEnabled) {
+                externalId = ExternalId.generate();
+            }
+
+            HashMap<String, BigDecimal> cumulativeIncomeFromInstallments = new HashMap<>();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: The two logic is not matching!
Things that got extracted from 3493 to 3508 is not matching with the new logic of the "applyIncomeAccrualTransaction" method. Please double check and remove the unnecessary pieces!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 4]
```diff
@@ -3801,6 +3822,8 @@ public ChangedTransactionDetail closeAsWrittenOff(final JsonCommand command, fin
 
         validateAccountStatus(LoanEvent.WRITE_OFF_OUTSTANDING);
 
+        final LocalDate writtenOffOnLocalDate = command.localDateValueOfParameterNamed(TRANSACTION_DATE);
+        this.closedOnDate = writtenOffOnLocalDate;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3834]: Why to move these dates?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 5]
```diff
@@ -3969,14 +3990,13 @@ public ChangedTransactionDetail close(final JsonCommand command, final LoanLifec
             final Money totalLoanOverpayment = calculateTotalOverpayment();
             if (totalLoanOverpayment.isGreaterThanZero() && getInArrearsTolerance().isGreaterThanOrEqualTo(totalLoanOverpayment)) {
                 // TODO - KW - technically should set somewhere that this loan
-                // has
-                // 'overpaid' amount
+                // has 'overpaid' amount
+                this.closedOnDate = closureDate;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:4002]: Why to move this date?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments and please dont forget to cover the new behaviour with proper tests!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3450]: I needed to move the set of the `closedOnDate` value before the Life cycle machine method call because when the event listener was called It was in null at that moment

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3834]: I needed to move the set of the `closedOnDate` value before the Life cycle machine method call because when the event listener was called It was in null at that moment

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3450]: Please move the actualMaturityDate as well then...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:4002]: I needed to move the set of the `closedOnDate` value before the Life cycle machine method call because when the event listener was called It was in null at that moment

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3450]: What is weird...even the loan got changed later, since it is the same instance, the updated field should be reflected there as well.

File: fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3450]: Done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Done

File: fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java]: Done It was moved

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: The logic is there

File: fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java [hunk 6]
```diff
@@ -345,4 +345,5 @@ private Set<Long> getNotifiableUserIds(Long officeId, String permission) {
         Collection<AppUser> usersWithPermission = users.stream().filter(aU -> aU.hasAnyPermission(permission, "ALL_FUNCTIONS")).toList();
         return usersWithPermission.stream().map(AppUser::getId).collect(toSet());
     }
+
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java]: Please remove this extra line

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 7]
```diff
@@ -3451,9 +3451,9 @@ private void handleLoanRepaymentInFull(final LocalDate transactionDate, final Lo
             }
         }
         if (isAllChargesPaid) {
+            this.closedOnDate = transactionDate;
             loanLifecycleStateMachine.transition(LoanEvent.REPAID_IN_FULL, this);
 
-            this.closedOnDate = transactionDate;
             this.actualMaturityDate = transactionDate;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Please move the actualMaturityDate as well...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 8]
```diff
@@ -3510,6 +3510,39 @@ private void processIncomeAccrualTransactionOnLoanClosure() {
...
+            BigDecimal fee = (BigDecimal) feeDetails.get(FEE);
+            if (fee == null) {
+                fee = BigDecimal.ZERO;
+            }
+            BigDecimal penalty = (BigDecimal) feeDetails.get(PENALTY);
+            if (penalty == null) {
+                penalty = BigDecimal.ZERO;
+            }
+            BigDecimal total = fee.add(penalty);
+            // TODO: calculate interest?
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3532]: Please check whether we need the interest calculation as well, or do it later...

File: fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java (hunk 6 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java]: Done

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java [hunk 9]
```diff
@@ -233,6 +236,109 @@ public void testApplyAndWaiveLoanSpecificDueDatePenaltyWithDisbursementDate() {
...
+        log.info("Operation date {}", transactionDate);
+        payloadJSON = LoanTransactionHelper.getSpecifiedDueDateChargesForLoanAsJSON(chargeId.toString(), operationDate, feeAmount);
+        postLoansLoanIdChargesResponse = loanTransactionHelper.addChargeForLoan(loanId, payloadJSON, responseSpec);
+        assertNotNull(postLoansLoanIdChargesResponse);
+        final Long loanChargeId03 = postLoansLoanIdChargesResponse.getResourceId();
+        assertNotNull(loanChargeId03);
+
+        getLoansLoanIdResponse = loanTransactionHelper.getLoan(requestSpec, responseSpec, loanId);
+        validateLoanAccount(getLoansLoanIdResponse, Double.valueOf(principalAmount), Double.valueOf("40.00"), true);
+        loanTransactionHelper.evaluateLoanTransactionData(getLoansLoanIdResponse, "loanTransactionType.accrual", Double.valueOf("50.00"));
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java]: Why is it 50? There was no periodic accrual execution yet...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 8 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:3532]: We have comment this will be later

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java (hunk 9 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java]: In my opinion and as I have seen in other cases the `runaccrual` call runs separately and when it is executed in the test, we already have passed the add the third Loan Charge

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 7 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Done!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java (hunk 9 above)
Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java]: In that case this test is not sufficient, because the add periodic accrual job will calculate the Accrual transaction and not the event of loan got closed. We need a test which makes sure, the last accrual transaction was created by not a job, but by the system when the loan got closed.

Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java]: Would you also please check whether the very last accrual transaction amount, penalty portion, transaction date is correct and it created the journal entries for it as well?

Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanChargeSpecificDueDateTest.java]: Fixed, there were some issues in this point, as we saw the IT was not using Accrual accounting and in the code the Fee Details was adding the PENALTIES and the read for the Accrual Transaction was reading PENALTY. That includes the old code too.

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResource.java [hunk 10]
```diff
@@ -166,7 +166,7 @@ public String retrieveAll(@Context final UriInfo uriInfo,
             + "journalentries/1?fields=officeName,glAccountId,entryType,amount\n" + "\n" + "journalentries/1?runningBalance=true\n" + "\n"
             + "journalentries/1?transactionDetails=true")
     @ApiResponses({
-            @ApiResponse(responseCode = "200", description = "OK", content = @Content(schema = @Schema(implementation = JournalEntryData.class))) })
+            @ApiResponse(responseCode = "200", description = "OK", content = @Content(schema = @Schema(implementation = JournalEntriesApiResourceSwagger.GetJournalEntriesTransactionIdResponse.class))) })
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResource.java]: This looks not right to me... This API is not returning the response with paging. I think the one you are looking for is the `retrieveAll` method.

I reckon for this method the right swagger class would be `JournalEntriesApiResourceSwagger.GetJournalEntriesTransactionIdResponse.JournalEntryTransactionPageItems.class`

However I would rename the `JournalEntryTransactionPageItems` to something more generic like `JournalEntryTransactionItem`

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResourceSwagger.java [hunk 11]
```diff
@@ -62,4 +64,77 @@ private PostJournalEntriesTransactionIdResponse() {
...
+            @Schema(example = "2")
+            public Long transactionId;
+            public EnumOptionType transactionType;
+        }
+
+        static final class JournalEntryTransactionPageItems {
+
+            private JournalEntryTransactionPageItems() {}
+
+            @Schema(example = "1")
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResourceSwagger.java]: Please add all the relevant fields from the JournalEntryData

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my comments!

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResource.java (hunk 10 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResource.java]: When you get the Journal entries using the Transaction Id, the current endpoint return an array (usually two items) with the gl journal entries but It returns with paging 

Screenshot 2023-02-02 at 12 56 42

Done, I've renamed the class

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResourceSwagger.java (hunk 11 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResourceSwagger.java]: Done, the other fields are used for the bulk import process

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResource.java (hunk 10 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResource.java]: You are still annotating the wrong method. The "retrieveAll" is called not the "retrieveJournalEntryById"

Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/api/JournalEntriesApiResource.java]: Done, updated both annotations
